test/unit/org/apache/cassandra/db/marshal/AbstractTypeTest.java

```diff
@@ -740,4 +801,927 @@ public String toString()
+
+        DeserializationHelper leftHelper = new DeserializationHelper(leftTable, MessagingService.current_version, DeserializationHelper.Flag.LOCAL, ColumnFilter.all(leftTable));
+        SerializationHelper rightHelper = new SerializationHelper(rightHeader);
+
+        qt().withExamples(10).forAll(rightGen).checkAssert(v -> {
+            Assertions.assertThatNoException().describedAs(typeRelDesc(".decompose", left, right)).isThrownBy(() -> {
+
+                // value compatibility means that we can use left's type serializer to decompose a value of right's type
+                ByteBuffer rightDecomposed = right.decompose((R) v);
+                L leftComposed = (L) left.compose(rightDecomposed);
+                ByteBuffer leftDecomposed = left.decompose(leftComposed);
+                assertThat(leftDecomposed.hasRemaining()).isEqualTo(rightDecomposed.hasRemaining());
+            });
+
+            Assertions.assertThatNoException().describedAs(typeRelDesc(".deserialize", left, right)).isThrownBy(() -> {
+                // serialization compatibility means that we can read a cell written using right's type serializer with left's type serializer;
+                // this additinoally imposes the requirement for storing the buffer lenght in the serialized form if the value is of variable length
+                // as well as, either both types serialize into a single or multiple cells
+                if (left.isSerializationCompatibleWith(right))
+                {
+                    if (!left.isMultiCell() && !right.isMultiCell())
+                        verifySerializationCompatibilityForSimpleCells(left, right, v, rightTable, rightColumn, rightHelper, leftHeader, leftHelper, leftColumn);
+                    else
+                        verifySerializationCompatibilityForComplexCells(left, right, v, rightTable, rightColumn, rightHelper, leftHeader, leftHelper, leftColumn);
+                }
+            });
+        });
+
+        if (!left.isCompatibleWith(right) || right.comparisonType == AbstractType.ComparisonType.NOT_COMPARABLE || left.comparisonType == AbstractType.ComparisonType.NOT_COMPARABLE)
+            return;
+
+        // types compatibility means that we can compare values of right's type using left's type comparator additionally
+        // to types being serialization compatible
+        qt().withExamples(10).forAll(rightGen, rightGen).checkAssert((r1, r2) -> {
+            Assertions.assertThatNoException().describedAs(typeRelDesc(".compare", left, right)).isThrownBy(() -> {
+                if (!left.isMultiCell() && !right.isMultiCell())
+                {
+                    // make sure that frozen<left> isCompatibleWith frozen<right> ==> left isCompatibleWith right
+                    assertThat(left.unfreeze().isCompatibleWith(right.unfreeze())).isTrue();
+                }
```

Comment:
can we pull this out of the property check?  the property check runs 10 times, where as this condition is unchanged, so can run *before*